File: shared/signal-codec.ts
/**
 * 信号编解码、模拟值与越界判断（前后端共用的唯一算法实现说明）。
 *
 * 本文件是该算法的 TypeScript 参考实现；后端 Java 实现在
 * com.canbus.signal.SignalCodec，与这里逐函数镜像，保证同一份定义
 * 在页面本地与系统后端产出相同结果。
 *
 * 约定：
 * - Intel（小端、@1）位布局：每个字节内部 LSB 在前，起始位 = 起始字节 * 8；
 * - 物理值 = 原始值 * factor + offset；
 * - 模拟值落在各信号 mock 区间内（integer 时取整），一定在量程内。
 */
import { catalog, getMessage, signalsOf, type MessageDef, type SignalDef } from './catalog';

/** 位布局所需的最小信号结构（DBC 文本解析出的信号也满足）。 */
interface BitLayout {
  startBit: number;
  bitLength: number;
}

/** 生成一个信号的模拟物理值。 */
export function mockValue(signal: SignalDef, random: () => number = Math.random): number {
  const value = signal.mock.min + random() * (signal.mock.max - signal.mock.min);
  return signal.mock.integer ? Math.floor(value) : value;
}

/** 按唯一定义随机挑一条消息（与收拢前一致：按消息出现顺序均匀选取）。 */
export function pickMessage(random: () => number = Math.random): MessageDef {
  return catalog.messages[Math.floor(random() * catalog.messages.length)];
}

/** 物理值 -> 原始整数值。 */
export function encodeRaw(signal: { factor: number; offset: number }, physical: number): number {
  return Math.round((physical - signal.offset) / signal.factor);
}

/** 原始整数 -> 物理值。 */
export function decodeRaw(signal: { factor: number; offset: number }, raw: number): number {
  return raw * signal.factor + signal.offset;
}

/** 把一个信号的原始值按 Intel 位布局写入字节数组。 */
export function writeBits(bytes: number[], signal: BitLayout, raw: number): void {
  for (let i = 0; i < signal.bitLength; i++) {
    const bitIndex = signal.startBit + i;
    if (bitIndex < bytes.length * 8) {
      const byteIndex = bitIndex >> 3;
      const bitInByte = bitIndex & 7;
      bytes[byteIndex] = (bytes[byteIndex] & ~(1 << bitInByte)) | (((raw >> i) & 1) << bitInByte);
    }
  }
}

/** 从字节数组中按 Intel 位布局取出一个信号的原始值。 */
export function readBits(bytes: number[], signal: BitLayout): number {
  let raw = 0;
  for (let i = 0; i < signal.bitLength; i++) {
    const bitIndex = signal.startBit + i;
    if (bitIndex < bytes.length * 8) {
      raw |= ((bytes[bitIndex >> 3] >> (bitIndex & 7)) & 1) << i;
    }
  }
  return raw;
}

/** 十六进制字符串（形如 "0A 1B"）-> 字节数组。 */
export function parseHex(data: string): number[] {
  const hex = data.replace(/\s/g, '');
  const bytes: number[] = [];
  for (let i = 0; i < hex.length; i += 2) {
    bytes.push(parseInt(hex.substring(i, i + 2), 16));
  }
  return bytes;
}

function toHexByte(value: number): string {
  return (value & 0xff).toString(16).padStart(2, '0').toUpperCase();
}

/** 字节数组 -> 大写、空格分隔的十六进制字符串。 */
export function formatHex(bytes: number[]): string {
  return bytes.map(toHexByte).join(' ');
}

export interface MockFramePayload {
  arbitrationId: number;
  dlc: number;
  data: string;
  decoded: Record<string, number>;
}

/**
 * 造一帧模拟数据：按唯一定义生成物理值、编码进 8 字节，并立即用同一套
 * 定义解码回填 decoded（页面与后端走完全相同的路径）。
 */
export function generateMockPayload(
  message: MessageDef = pickMessage(),
  random: () => number = Math.random
): MockFramePayload {
  const signals = signalsOf(message);
  const bytes = new Array(catalog.frameLength).fill(0);

  for (const signal of signals) {
    writeBits(bytes, signal, encodeRaw(signal, mockValue(signal, random)));
  }

  return {
    arbitrationId: message.id,
    dlc: catalog.frameLength,
    data: formatHex(bytes),
    decoded: decodeMessageBytes(bytes, message)
  };
}

/** 可解码的信号结构（唯一定义与 DBC 文本解析结果都满足）。 */
export interface DecodableSignal {
  name: string;
  startBit: number;
  bitLength: number;
  factor: number;
  offset: number;
}

/** 按信号列表解码原始字节，返回 信号名 -> 物理值（保留定义给出的全精度）。 */
export function decodeBytes(
  bytes: number[],
  signals: ReadonlyArray<DecodableSignal>
): Record<string, number> {
  const decoded: Record<string, number> = {};
  for (const signal of signals) {
    decoded[signal.name] = decodeRaw(signal, readBits(bytes, signal));
  }
  return decoded;
}

/** 按唯一定义中的消息解码原始字节。 */
export function decodeMessageBytes(bytes: number[], message: MessageDef): Record<string, number> {
  return decodeBytes(bytes, signalsOf(message));
}

/** 按报文 ID 解码十六进制数据串；没有该消息定义时返回 null。 */
export function decodeDataHex(arbitrationId: number, data: string): Record<string, number> | null {
  const message = getMessage(arbitrationId);
  if (!message) return null;
  return decodeMessageBytes(parseHex(data), message);
}

/** 越界判断：值严格落在定义的量程闭区间之外即为越界；未知信号不判越界。 */
export function isOutOfRange(signal: { minValue: number; maxValue: number }, value: number): boolean {
  return value < signal.minValue || value > signal.maxValue;
}

/**
 * 量程百分比（用于详情条宽度）：把值映射到 [min,max] 对应 [0,100]，
 * 越界部分钳制在 0–100，与收拢前展示表现一致。
 */
export function rangePercent(
  signal: { minValue: number; maxValue: number },
  value: number
): number {
  if (signal.maxValue === signal.minValue) return 0;
  const percent = ((value - signal.minValue) / (signal.maxValue - signal.minValue)) * 100;
  return Math.max(0, Math.min(100, percent));
}

File: backend/src/main/java/com/canbus/signal/SignalCodec.java
package com.canbus.signal;

import org.springframework.stereotype.Component;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Random;

/**
 * 模拟值生成与报文编解码的后端实现，与前端 shared/signal-codec.ts 逐函数镜像：
 * <ul>
 *   <li>Intel（小端）位布局：起始位 = 起始字节 * 8，每字节内 LSB 在前；</li>
 *   <li>物理值 = 原始值 * factor + offset；</li>
 *   <li>按唯一定义造模拟值并编码进字节，再用同一套定义解码回填。</li>
 * </ul>
 * 因为读取的是同一份 signals.json，前后端对任意帧得到相同解码结果。
 */
@Component
public class SignalCodec {

    private final SignalDefs defs;

    public SignalCodec(SignalDefs defs) {
        this.defs = defs;
    }

    /** 物理值 -> 原始整数。 */
    public int encodeRaw(SignalDef signal, double physical) {
        return (int) Math.round((physical - signal.offset()) / signal.factor());
    }

    /** 原始整数 -> 物理值（全精度）。 */
    public double decodeRaw(SignalDef signal, int raw) {
        return raw * signal.factor() + signal.offset();
    }

    /** 把一个信号的原始值按 Intel 位布局写入字节数组。 */
    public void writeBits(int[] bytes, SignalDef signal, int raw) {
        for (int i = 0; i < signal.bitLength(); i++) {
            int bitIndex = signal.startBit() + i;
            if (bitIndex < bytes.length * 8) {
                int byteIndex = bitIndex >> 3;
                int bitInByte = bitIndex & 7;
                bytes[byteIndex] = (bytes[byteIndex] & ~(1 << bitInByte))
                        | (((raw >> i) & 1) << bitInByte);
            }
        }
    }

    /** 从字节数组按 Intel 位布局取出一个信号的原始值。 */
    public int readBits(int[] bytes, SignalDef signal) {
        int raw = 0;
        for (int i = 0; i < signal.bitLength(); i++) {
            int bitIndex = signal.startBit() + i;
            if (bitIndex < bytes.length * 8) {
                raw |= ((bytes[bitIndex >> 3] >> (bitIndex & 7)) & 1) << i;
            }
        }
        return raw;
    }

    /** 十六进制字符串（形如 "0A 1B"）-> 字节数组。 */
    public int[] parseHex(String data) {
        String hex = data.replaceAll("\\s", "");
        int[] bytes = new int[hex.length() / 2];
        for (int i = 0; i < bytes.length; i++) {
            bytes[i] = Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return bytes;
    }

    /** 字节数组 -> 大写、空格分隔的十六进制字符串。 */
    public String formatHex(int[] bytes) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < bytes.length; i++) {
            if (i > 0) {
                builder.append(' ');
            }
            builder.append(String.format("%02X", bytes[i] & 0xFF));
        }
        return builder.toString();
    }

    /** 按消息定义解码字节，保留定义给出的全精度。 */
    public Map<String, Double> decodeBytes(int[] bytes, MessageDef message) {
        Map<String, Double> decoded = new LinkedHashMap<>();
        for (SignalDef signal : defs.signalsOf(message)) {
            decoded.put(signal.name(), decodeRaw(signal, readBits(bytes, signal)));
        }
        return decoded;
    }

    /** 报文接口读数写法：四舍五入保留两位小数（与收拢前后端一致）。 */
    public Map<String, Double> decodeForApi(int[] bytes, MessageDef message) {
        Map<String, Double> decoded = new LinkedHashMap<>();
        for (Map.Entry<String, Double> entry : decodeBytes(bytes, message).entrySet()) {
            decoded.put(entry.getKey(), Math.round(entry.getValue() * 100.0) / 100.0);
        }
        return decoded;
    }

    /**
     * 造一帧模拟数据：物理值生成、字节编码、解码回填全部来自唯一定义。
     * 返回 [hex 数据串, 解码读数]，arbId/dlc 由调用方按消息写入 CanFrame。
     */
    public MockPayload generateMockPayload(MessageDef message, Random random) {
        int[] bytes = new int[defs.catalog().frameLength()];

        for (SignalDef signal : defs.signalsOf(message)) {
            writeBits(bytes, signal, encodeRaw(signal, signal.mockValue(random)));
        }

        return new MockPayload(formatHex(bytes), decodeForApi(bytes, message));
    }

    /** 按定义均匀随机挑选一条消息（与前端 pickMessage 镜像）。 */
    public MessageDef pickMessage(Random random) {
        var messages = defs.messages();
        return messages.get(random.nextInt(messages.size()));
    }

    public record MockPayload(String data, Map<String, Double> decoded) {
    }
}
